File: NeetCode/Stack/Easy/20.ValidParantheses.py
# Given a string s containing just the characters 
# '(', ')', '{', '}', '[' and ']', determine if the input string is valid.

def isValid(s):
    stack = []
    dict = {")": "(", "]": "[", "}": "{"}

    for item in s:
        if item in dict:
            if stack and stack[-1] == dict[item]:
                stack.pop()
            else:
                return False
        else:
            stack.append(item)

    return True if not stack else False

print(isValid("()"))
print(isValid("()[]{}"))
print(isValid("(]"))

File: swift/Neetcode/Stack/Easy/20.ValidParantheses.swift
// Given a string s containing just the characters 
// '(', ')', '{', '}', '[' and ']', determine if the input string is valid.

// Time: O(n), Space: O(n)
func isValid(_ s: String) -> Bool {
    var dict: [Character: Character] = [")": "(", "]": "[", "}": "{"]
    var stack = [Character]()

    for letter in s {
        if !dict.keys.contains(letter) {
            stack.append(letter)
        } else if let last = stack.last, dict[letter]! == last {
            stack.removeLast()
        } else {
            return false
        }
    }

    return stack.isEmpty
}

print(isValid("()"))
print(isValid("()[]{}"))
print(isValid("(]"))